chaosreliably/controls/experiment.py

```diff
@@ -0,0 +1,310 @@
+) -> None:
+    """
+    Control run *before* the execution of an Experiment.
+
+    For a given Experiment, the control creates (if not already created) an Experiment
+    Entity Context and an Experiment Version Entity Context in the Reliably Entity
+    Server.
+
+    A unique Experiment Run Entity Context is also created, with an Experiment Event
+    Entity Context of type `EXPERIMENT_START` created, relating to the run.
+
+    The control requires the `arguments` of `commit_hash`, `source`, and `user` to be
+    provided to the control definition. If not provided, the control will simply not
+    create any Entity Contexts.
+
+    Once the Entity Contexts have been created, an entry into the configuration is made
+    under configuration["chaosreliably"]["experiment_run_labels"] to allow for
+    other controls to create events relating to the Experiment Run.
+
+    :param context: Experiment object representing the Experiment that will be executed
+    :param configuration: Configuration object provided by Chaos Toolkit
+    :param secrets: Secret object provided by Chaos Toolkit
+    :param **kwargs: Expected `kwargs` are 'commit_hash' (str), `source` (str), and
+        `user` (str)
+
+    Examples
+    --------
+
+    "controls": [
+        {
+            "name": "chaosreliably",
+            "provider": {
+                "type": "python",
+                "module": "chaosreliably.controls"
+                "arguments": {
+                    "commit_hash": "59f9f577e2d90719098f4d23d26329ce41f2d0bd",
+                    "source": "https://github.com/chaostoolkit-incubator/chaostoolkit-reliably/exp.json",  # Noqa
+                    "user": "A users name"
+                }
+            }
+        }
+    ]
+    """
+    try:
+        commit_hash = kwargs.get("commit_hash")
+        source = kwargs.get("source")
+        user = kwargs.get("user")
+        if not commit_hash or not source or not user:
+            logger.warn(
+                "The parameters: `commit_hash`, `source`, and `user` are required for "
+                "the chaosreliably controls, please provide them. This Experiment Run"
+                " will not be tracked with Reliably."
+            )
+            return
+
+        experiment_run_labels = (
+            _create_experiment_entities_for_before_experiment_control(
+                experiment_title=context["title"],
+                commit_hash=commit_hash,
+                source=source,
+                user=user,
+                configuration=configuration,
+                secrets=secrets,
+            )
+        )
+
+        configuration.update(
+            {"chaosreliably": {"experiment_run_labels": experiment_run_labels.dict()}}
+        )
+    except Exception as ex:
+        logger.warn(
+            f"An error occurred: {ex}, whilst running the Before Experiment control, "
+            "no further entities will be created, the Experiment execution won't be "
+            "affected"
+        )
```

Comment:
Failures are caught by the control callers, not sure you need to capture them here. That said, the message is more specific your way of course.

Reply:
But in this case, I would suggest you give details about what the user should do to fix this.